qemu/tests/win_virtio_driver_install_from_update.py

```diff
@@ -0,0 +1,91 @@
+import logging
+import time
+
+from virttest import error_context
+from virttest import utils_misc
+
+from provider import win_driver_utils
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Install drivers from windows update:
+
+    1) Boot guest with the device.
+    2) Check windows update service is running
+       if not, start the service.
+    3) Install drivers from windows update.
+    4) Run driver signature check command in guest.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def start_wuauserv_service(session):
+        """
+        Start wuauserv service.
+
+        :param session: The guest session object.
+        """
+        wuauserv_status_cmd = params["wuauserv_status_cmd"]
+        status = session.cmd_status(wuauserv_status_cmd)
+        if status != 0:
+            wuauserv_service_cfg_cmd = params["wuauserv_service_cfg_cmd"]
+            status = session.cmd_status(wuauserv_service_cfg_cmd)
+            if status != 0:
+                test.fail("Change wuauserv service config not success")
+            wuauserv_start_cmd = params["wuauserv_start_cmd"]
+            status = session.cmd_status(wuauserv_start_cmd)
+            if status != 0:
+                test.fail("Fail to start wuauserv service")
+
+    driver_name = params["driver_name"]
+    device_name = params["device_name"]
+    device_hwid = params["device_hwid"]
+    install_driver_cmd = params["install_driver_cmd"]
+    check_stat = params["check_driver_stat"] % driver_name
+    chk_cmd = params["vio_driver_chk_cmd"] % device_name[0:30]
+    chk_timeout = int(params.get("chk_timeout", 240))
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+
+    error_context.context("Start wuauserv service", logging.info)
+    start_wuauserv_service(session)
+
+    error_context.context("Uninstall %s driver" % driver_name, logging.info)
+    win_driver_utils.uninstall_driver(session, params, driver_name,
+                                      device_name, device_hwid)
+    session = vm.reboot(session)
+
+    error_context.context("Install drivers from windows update", logging.info)
+    install_driver_cmd = utils_misc.set_winutils_letter(session,
+                                                        install_driver_cmd)
+    vm.send_key('meta_l-d')
+    time.sleep(30)
+    session.cmd(install_driver_cmd)
+    # workaround for viostor and vioscsi as driver status still be running
+    # after uninstall
+    if driver_name in ('viostor', 'vioscsi'):
+        time.sleep(120)
```

Comment:
Better check it before install the new driver instead of simple sleep here.

Reply:
time.sleep(120) should be necessary here to be a workaround, for we should wait some time to let autoit script  session.cmd(install_driver_cmd)   to install driver from windows update.